File: src/main/java/frc/robot/subsystems/superstructure/SuperstructureGoals.java
package frc.robot.subsystems.superstructure;

import com.pathplanner.lib.path.PathConstraints;
import edu.wpi.first.math.geometry.Rotation2d;
import frc.robot.subsystems.drive.DriveConstants;

/** Predefined superstructure targets. */
public final class SuperstructureGoals {
    private SuperstructureGoals() {
    }

    public static final SuperstructureState STOW = new SuperstructureState(Rotation2d.fromDegrees(63.0),
            DriveConstants.DEFAULT_LIMITS, null, "Stow");

    public static final SuperstructureState PRE_ALGAE_L2 = new SuperstructureState(
            Rotation2d.fromDegrees(38.0), DriveConstants.EXTENDED_LIMITS, null, "Pre Algae L2");

    public static final SuperstructureState PRE_ALGAE_L3 = new SuperstructureState(
            Rotation2d.fromDegrees(125.0), DriveConstants.EXTENDED_LIMITS, null, "Pre Algae L3");

    public static final SuperstructureState ALGAE_L2 = new SuperstructureState(
            Rotation2d.fromDegrees(38.0), DriveConstants.EXTENDED_LIMITS, null, "Algae L2");

    public static final SuperstructureState ALGAE_L3 = new SuperstructureState(
            Rotation2d.fromDegrees(125.0), DriveConstants.EXTENDED_LIMITS, null, "Algae L3");

    public static final SuperstructureState L1 = new SuperstructureState(Rotation2d.fromDegrees(34.0),
            DriveConstants.EXTENDED_LIMITS, null, "L1");

    public static final SuperstructureState GROUND = new SuperstructureState(Rotation2d.fromDegrees(0.0),
            DriveConstants.EXTENDED_LIMITS, null, "Ground");

    public static final SuperstructureState PROCESSOR = new SuperstructureState(
            Rotation2d.fromDegrees(35.0), DriveConstants.EXTENDED_LIMITS, null, "Processor");

    public static final SuperstructureState ALGAE_INTAKE = new SuperstructureState(
            Rotation2d.fromDegrees(30.0), DriveConstants.EXTENDED_LIMITS, null, "Algae Intake");
}

File: src/main/java/frc/robot/subsystems/superstructure/SuperstructureGoals.kt
package frc.robot.subsystems.superstructure

import edu.wpi.first.units.Units.*
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.Distance
import edu.wpi.first.util.struct.Struct
import edu.wpi.first.util.struct.StructSerializable
import frc.robot.subsystems.swerve.Drivebase
import lib.math.units.into
import java.nio.ByteBuffer
import java.util.Optional

// TODO: these need to be tuned significantly
object SuperstructureGoals {
    val STOW = SuperstructureState(
        armAngle = Degrees.of(63.0),
        driveLimits = Drivebase.defaultLimits,
        name = "Stow"
    )

    // val ALGAE_STOW = SuperstructureState(
    //     armAngle = Degrees.of(-20.0),
    //     driveLimits = Drivebase.defaultLimits,
    //     name = "Algae Stow"
    // )

    val PRE_ALGAE_L2 = SuperstructureState(
        armAngle = Degrees.of(63.0),
        driveLimits = Drivebase.extendedLimits,
        name = "Pre Algae L2"
    )

    val PRE_ALGAE_L3 = SuperstructureState(
        armAngle = Degrees.of(125.0),
        driveLimits = Drivebase.extendedLimits,
        name = "Pre Algae L3"
    )

    val ALGAE_L2 = SuperstructureState(
        armAngle = Degrees.of(34.0),
        driveLimits = Drivebase.extendedLimits,
        name = "Algae L2"
    )

    val ALGAE_L3 = SuperstructureState(
        armAngle = Degrees.of(125.0),
        driveLimits = Drivebase.extendedLimits,
        name = "Algae L3"
    )

    // val BARGE = SuperstructureState(
    //     armAngle = Degrees.of(-20.0),
    //     driveLimits = Drivebase.extendedLimits,
    //     name = "Barge"
    // )
    
    val L1 = SuperstructureState(
        armAngle = Degrees.of(34.0), // needs tuning
        driveLimits = Drivebase.extendedLimits,
        name = "L1"
    )

    val GROUND = SuperstructureState(
        armAngle = Degrees.of(1.0),
        driveLimits = Drivebase.extendedLimits,
        name = "Ground"
    )
    
    // val L2 = SuperstructureState(
    //     armAngle = scoreAngle,
    //     driveLimits = Drivebase.extendedLimits,
    //     name = "L2"
    // )
    
    // val L3 = SuperstructureState(
    //     armAngle = scoreAngle,
    //     driveLimits = Drivebase.extendedLimits,
    //     name = "L3"
    // )
    
    // val L4 = SuperstructureState(
    //     armAngle = Degrees.of(20.0),
    //     driveLimits = Drivebase.extendedLimits,
    //     name = "L4"
    // )
}